### PR TITLE
feat(monaco-esm-build): utilize Monaco esm classes

### PR DESCRIPTION
**Pull request by thomaslennan**
## Description

Utilized custom electron webpack rules to import esm version of Monaco classes.

Code wise, stripped out all the inner html code. Modified the code-editor component to reduce the amount of editor rapid destruction/re-construction events.

Addressed issue with highlight component not detecting language change, but backed out the changes. Changes to highlight will be addressed under https://github.com/Teradata/covalent/issues/1761.

### What's included?

- Custom webpack
- Changes for code-editor component
- Tweaked spec file to allow time for editor value/language to be set.

#### Test Steps
- [ ] clone and npm ci
- [ ] `npm run serve`
- [ ] Navigate to Components/code-editor
- [ ] Open developer tools and show console
- [ ] Change language, confirm that no errors show up in the console
- [ ] Change editor content, confirm no errors and that changes are reflected in markup view

- [ ] Run tests

#### General Tests for Every PR

- [ ] `npm run serve:prod` still works.
- [ ] `npm run tslint` passes.
- [ ] `npm run stylelint` passes.
- [ ] `npm test` passes and code coverage is not lower.
- [ ] `npm run build:lib` still works.

### CODE REVIEW

File: angular.json [hunk 1]
```diff
@@ -211,8 +237,14 @@
           }
         },
         "test": {
-          "builder": "@angular-devkit/build-angular:karma",
```
Comment (christianmemije) [inline angular.json:214]: I think we should set up the custom webpack config in the build config and then reference the browserTarget in the test config.

File: package.json [hunk 2]
```diff
@@ -86,7 +86,7 @@
     "echarts-wordcloud": "^1.1.3",
     "hammerjs": "^2.0.8",
     "highlight.js": "9.13.1",
-    "monaco-editor": "^0.17.0",
+    "monaco-editor": "^0.20.0",
```
Comment (christianmemije) [inline package.json:89]: Make sure to update build.conf.js

File: package.json [hunk 3]
```diff
@@ -95,6 +95,7 @@
     "zone.js": "^0.10.2"
   },
   "devDependencies": {
+    "@angular-builders/custom-webpack": "8.4.1",
```
Comment (christianmemije) [inline package.json:98]: Let's use carets ^ for dependencies

File: src/platform/code-editor/README.md [hunk 4]
```diff
@@ -48,21 +48,68 @@ npm install @covalent/code-editor
...
+npm install --save-dev monaco-editor-webpack-plugin
+```
+
+Create a webpack config file utilizing the Monaco Editor webpack plugin. Languages and features can be included/excluded to control the resulting image size.
+
+```javascript
+const MonacoWebpackPlugin = require('monaco-editor-webpack-plugin');
+
+module.exports = {
+  target: 'electron-renderer',
```
Comment (christianmemije) [inline src/platform/code-editor/README.md:71]: I would add an inline comment here explaining this is optional

File: src/platform/code-editor/code-editor.component.ts [hunk 5]
```diff
@@ -41,31 +36,29 @@ declare const monaco: any;
...
-  private _componentInitialized: boolean = false;
-  private _fromEditor: boolean = false;
-  private _editorOptions: any = {};
-  private _isFullScreen: boolean = false;
-  private _keycode: any;
-  private _setValueTimeout: any;
-  private initialContentChange: boolean = true;
-  private _registeredLanguagesStyles: HTMLStyleElement[] = [];
+export class TdCodeEditorComponent implements AfterViewInit, ControlValueAccessor, OnDestroy {
+  _destroy: Subject<boolean> = new Subject<boolean>();
```
Comment (christianmemije) [inline src/platform/code-editor/code-editor.component.ts:40]: Why changed them to be public?

File: src/platform/code-editor/code-editor.component.ts [hunk 6]
```diff
@@ -101,66 +94,30 @@ export class TdCodeEditorComponent implements OnInit, AfterViewInit, ControlValu
...
+      // don't want to keep sending content if event came from the editor, infinite loop
+      if (!this._fromEditor) {
+        this._editor.setValue(value);
       }
+      this.editorValueChange.emit();
+      this.propagateChange(this._value);
+      this.change.emit();
+      this._fromEditor = false;
     } else {
       this._setValueTimeout = setTimeout(() => {
```
Comment (christianmemije) [inline src/platform/code-editor/code-editor.component.ts:122]: We shouldn't need timeouts anymore right? Since we change the way it's loaded.

File: src/platform/code-editor/code-editor.component.ts [hunk 7]
```diff
@@ -217,32 +169,43 @@ export class TdCodeEditorComponent implements OnInit, AfterViewInit, ControlValu
...
+        ),
+      );
+      this._editor.getModel().onDidChangeContent((e: any) => {
+        this._fromEditor = true;
+        this.writeValue(this._editor.getValue());
+      });
+      this.editorConfigurationChanged.emit();
+      this.editorLanguageChanged.emit();
+    } else {
+      this._setLanguageTimeout = setTimeout(() => {
```
Comment (christianmemije) [inline src/platform/code-editor/code-editor.component.ts:206]: We shouldn't need timeouts.

File: src/platform/code-editor/code-editor.component.ts [hunk 8]
```diff
@@ -778,7 +483,7 @@ export class TdCodeEditorComponent implements OnInit, AfterViewInit, ControlValu
   /**
    * addFullScreenModeCommand used to add the fullscreen option to the context menu
    */
-  private addFullScreenModeCommand(): void {
+  addFullScreenModeCommand(): void {
```
Comment (christianmemije) [inline src/platform/code-editor/code-editor.component.ts:486]: Why public?

File: src/platform/core/schematics/tsconfig.json [hunk 9]
```diff
@@ -10,7 +10,7 @@
     "lib": ["es2017", "dom"],
     "moduleResolution": "node",
     "sourceMap": true,
-    "target": "es5",
+    "target": "es2015",
```
Comment (christianmemije) [inline src/platform/core/schematics/tsconfig.json:13]: Why this change in the schematics tsconfig? Seems unrelated no?

File: src/platform/highlight/highlight.component.ts [hunk 10]
```diff
@@ -48,7 +49,13 @@ export class TdHighlightComponent implements AfterViewInit {
    *
    * e.g. `typescript`, `html` , etc.
    */
-  @Input() lang: string = 'typescript';
+  @Input('lang')
```
Comment (christianmemije) [inline src/platform/highlight/highlight.component.ts:52]: These highlight changes are separate from the editor changes. Can we revert them and open a separate PR?

Comment (thomaslennan) [reply src/platform/highlight/highlight.component.ts]: Reverted.

File: src/platform/core/schematics/tsconfig.json (hunk 9 above)
Comment (thomaslennan) [reply src/platform/core/schematics/tsconfig.json]: Reverted.

File: src/platform/code-editor/code-editor.component.ts (hunk 8 above)
Comment (thomaslennan) [reply src/platform/code-editor/code-editor.component.ts:415]: Reverted.

File: src/platform/code-editor/code-editor.component.ts (hunk 7 above)
Comment (thomaslennan) [reply src/platform/code-editor/code-editor.component.ts]: Removed.

File: src/platform/code-editor/code-editor.component.ts (hunk 6 above)
Comment (thomaslennan) [reply src/platform/code-editor/code-editor.component.ts]: Removed.

File: src/platform/code-editor/code-editor.component.ts (hunk 5 above)
Comment (thomaslennan) [reply src/platform/code-editor/code-editor.component.ts]: Reverted.

File: src/platform/code-editor/README.md (hunk 4 above)
Comment (thomaslennan) [reply src/platform/code-editor/README.md:72]: Done.

File: package.json (hunk 3 above)
Comment (thomaslennan) [reply package.json]: Done.

File: package.json (hunk 2 above)
Comment (thomaslennan) [reply package.json:89]: Done.

File: angular.json (hunk 1 above)
Comment (thomaslennan) [reply angular.json:214]: Not sure what you're asking for here.

File: src/platform/code-editor/code-editor.component.ts [hunk 11]
```diff
@@ -101,77 +89,31 @@ export class TdCodeEditorComponent implements OnInit, AfterViewInit, ControlValu
...
-          this._fromEditor = false;
-        } else {
-          // Editor is not loaded yet, try again in half a second
-          this._setValueTimeout = setTimeout(() => {
-            this.value = value;
-          }, 500);
-        }
-      } else {
-        if (this._editor && this._editor.setValue) {
-          // don't want to keep sending content if event came from the editor, infinite loop
```
Comment (christianmemije) [inline src/platform/code-editor/code-editor.component.ts:148]: Do we not have to account for this case where it comes from the editor?

File: src/platform/code-editor/code-editor.component.ts [hunk 12]
```diff
@@ -219,80 +156,59 @@ export class TdCodeEditorComponent implements OnInit, AfterViewInit, ControlValu
...
-              theme: this._theme,
-            },
-            this.editorOptions,
-          ),
-        );
-        this._editor.getModel().onDidChangeContent((e: any) => {
-          this._fromEditor = true;
-          this.writeValue(this._editor.getValue());
-        });
-        this.editorConfigurationChanged.emit();
```
Comment (christianmemije) [inline src/platform/code-editor/code-editor.component.ts:243]: Do we not have to emit thie `editorConfigurationChanged` event anymore?

File: src/platform/code-editor/code-editor.component.ts (hunk 11 above)
Comment (thomaslennan) [reply src/platform/code-editor/code-editor.component.ts:148]: Shouldn't be an issue since we're now checking if value requested to be set is current component value.

File: src/platform/code-editor/code-editor.component.ts (hunk 12 above)
Comment (thomaslennan) [reply src/platform/code-editor/code-editor.component.ts:243]: Reverted.

File: angular.json [hunk 13]
```diff
@@ -79,8 +80,14 @@
           }
         },
         "serve": {
-          "builder": "@angular-devkit/build-angular:dev-server",
+          "builder": "@angular-builders/custom-webpack:dev-server",
           "options": {
+            "customWebpackConfig": {
```
Comment (christianmemije) [inline angular.json:85]: Shouldn't need this since it is already in build

File: angular.json [hunk 14]
```diff
@@ -185,8 +199,14 @@
...
```
Comment (christianmemije) [inline angular.json:204]: Shouldn't need this since it is already in build

File: angular.json [hunk 15]
```diff
@@ -396,26 +416,32 @@
       "prefix": "lib",
       "architect": {
         "build": {
-          "builder": "@angular-devkit/build-ng-packagr:build",
+          "builder": "@angular-builders/custom-webpack:build",
```
Comment (christianmemije) [inline angular.json:419]: Should be `:browser`
